### PR TITLE
CD-242902 Added logs for token requests and cache key when it is empty

### PR DESCRIPTION
**Pull request by supriya**
* Added a configurable option to spit out extra logs if we need to. At the moment, those extra logs are being spit only at the token requests. 
* Logging the token request caller, truncating the trace up to 25, maybe we can make this number configurable too?

- [ ] @johnny-lai 
- [x] @abdulchaudhrycoupa 
- [x] @johnwu96822

### CODE REVIEW

File: lib/sand/client.rb [hunk 1]
```diff
@@ -131,6 +131,7 @@ def token_with_expiry(options = {})
           end
         end
       end
+      logger&.warn("Cache with key #{ckey} not set")
```
Comment (supriya) [inline lib/sand/client.rb:134]: not checking for the `@extra_logging_enabled` as ideally this should not occur often.

File: lib/sand/client.rb [hunk 2]
```diff
@@ -145,6 +146,7 @@ def token_with_expiry(options = {})
     # backoff time of 1, 2, 4... seconds
     # If options[:num_retry] is not present or < 0, it will use @default_retry_count as the retry number
     def oauth_token(options = {})
+      logger&.warn("#{Time.now.utc} Oauth token request call with process #{Process.pid} from: #{caller[0..25]}") if @extra_logging_enabled
```
Comment (supriya) [inline lib/sand/client.rb:149]: thinking if number `25` should also be configurable?

Comment (abdulchaudhrycoupa) [reply lib/sand/client.rb:149]: minor: maybe add Process.ppid as will

**Review by abdulchaudhrycoupa — APPROVED** [review]

File: lib/sand/client.rb (hunk 2 above)
Comment (johnwu96822) [reply lib/sand/client.rb:149]: `caller[0..25].join("\n")`

Comment (johnwu96822) [reply lib/sand/client.rb:149]: `if @extra_logging_enabled == true` if you want to enable verbose logging only if this value is true.

File: lib/sand/client.rb (hunk 1 above)
Comment (johnwu96822) [reply lib/sand/client.rb:134]: Can move this to line 133 https://github.com/coupa/sand-ruby/pull/32/files#diff-dc62044eda9e5629af9bdc59eb9259edfcb1e8ae45cc8c24a92b640645b5f21bR133

And change it to: `logger&.info("Not using the cache with #{ckey}")`

File: lib/sand/base.rb [hunk 3]
```diff
@@ -17,7 +17,7 @@ class Base
 
     attr_accessor :client_id, :client_secret, :token_site, :token_path,
         :ssl_min_version, :default_retry_count, :cache_root, :logger,
-        :min_token_valid_s, :timeout, :open_timeout
+        :min_token_valid_s, :timeout, :open_timeout, :extra_logging_enabled
```
[the remark this answers is not in the document]
Comment (johnny-lai) [reply lib/sand/base.rb:20]: `:verbose` has my vote

File: lib/sand/base.rb [hunk 4]
```diff
@@ -52,6 +52,7 @@ def initialize(opts = {})
       @timeout = (opts.delete(:timeout) || 5).to_f
       @open_timeout = (opts.delete(:open_timeout) || 5).to_f
       @ssl_min_version = opts.delete(:ssl_min_version)
+      @extra_logging_enabled = opts.delete(:extra_logging_enabled) || true
```
Comment (johnny-lai) [inline lib/sand/base.rb:55]: if extra_logging_enabled is set to `false`, I think this still ends up as `true`

File: lib/sand/client.rb [hunk 5]
```diff
@@ -145,6 +146,7 @@ def token_with_expiry(options = {})
     # backoff time of 1, 2, 4... seconds
     # If options[:num_retry] is not present or < 0, it will use @default_retry_count as the retry number
     def oauth_token(options = {})
+      logger&.warn("#{Time.now.utc} Oauth token request call with process #{Process.pid} and parent process #{Process.ppid} from:\n#{caller[0..25].join('\n')}") if @verbose == true
```
Comment (johnwu96822) [inline lib/sand/client.rb:149]: `logger&.info`

Comment (johnwu96822) [reply lib/sand/client.rb:149]: Perhaps we don't need `#{Time.now.utc}` any more? Since that may be included by the log entry.

**Review by johnwu96822 — APPROVED** [review]

**Review by johnny-lai — APPROVED** [review]

**Comment by supriya** [comment]
[lockit v4.9.3] merged via Lockit CLI on 2021-05-26 14:00:18.362338 -0700 PDT m=+4.628423496

**Pull request merged**